### PR TITLE
Online Transfer payment methods

### PR DESCRIPTION
**Pull request by aleksey-kotsuba**
- Developer(s): @aleksey-kotsuba

- Ticket: 
               https://spryker.atlassian.net/browse/ECO-1797
               https://spryker.atlassian.net/browse/ECO-1798
               https://spryker.atlassian.net/browse/ECO-1799
               https://spryker.atlassian.net/browse/ECO-1800
               https://spryker.atlassian.net/browse/ECO-1801
               https://spryker.atlassian.net/browse/ECO-1802
               https://spryker.atlassian.net/browse/ECO-1803
               https://spryker.atlassian.net/browse/ECO-1804

- PR&CR: https://spryker.atlassian.net/browse/ECO-1891

- Academy PR: https://spryker.atlassian.net/wiki/spaces/ECO/pages/473989505/Adyen+Documentation

#### Please confirm

- [ ] No new OS components - or they have been approved by legal department

#### Documentation

- [ ] Functional documentation provided or in progress?
- [ ] Integration guide for projects provided or not needed?
- [ ] Migration guides for all contained majors provided or not needed?

#### Release Table

   Module                | Release Type         | Constraint Updates         |
   :--------------------- | :------------------------ | :--------------------- |
   Adyen                 | major (new)           |                       |

#### Release Notes
* Added CreditCard with 3D secure payment method.
* Added Sofort payment method.
* Added DirectDebit payment method.
* Added Prepayment payment method.
* Added Klarna Invoice payment method.
* Added PayPal payment method.
* Added iDeal payment method.
* Added AliPay payment method.
* Added WeChatPay payment method.

-----------------------------------------

#### Module Adyen

_**Maxi-Major:** New or fully refactored module._

Def of done (by responsible developer):
- [ ] API methods are defined and have specifications [API-doc](https://academy.spryker.com/understanding_spryker/definition_api.html).
- [ ] All [dependencies](https://spryker.atlassian.net/wiki/spaces/CORE/pages/4980746/Module+Dependency+Guidelines) are defined, checked and corrected.
- [ ] Cleanups are resolved or closed. [cleanups](https://release.spryker.com/module-cleanups).
- [ ] New and changed facade methods covered by functional tests. / Has nothing to test.
- [ ] New and changed complex business logic is covered by unit or integration tests. / Has nothing to test.
- [ ] Migration guide provided / It's a new module.
- [ ] Module description in README is up to date.

### Improvements
* Next payment methods have been added:
** CC with 3D secure;
** Sofort;
** DirectDebit;
** Prepayment;
** Klarna Invoice;
** PayPal;
** iDeal;
** AliPay;
** WeChatPay;
* Module structure has been refactored and improved.

### CODE REVIEW

File: src/SprykerEco/Yves/Adyen/Theme/default/credit_card.twig [hunk 1]
```diff
@@ -76,4 +79,4 @@
         }
     });
 
-</script>
+</script>
```
Comment (limeeugenia) [inline src/SprykerEco/Yves/Adyen/Theme/default/credit_card.twig]: Newline missing.

File: src/SprykerEco/Yves/Adyen/Theme/default/direct_debit.twig [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+        {{ form_widget(form.adyenDirectDebit.owner_name, { 'attr': {'placeholder': 'adyenDirectDebit.owner_name' | trans } }) }}
+        {{ form_errors(form.adyenDirectDebit.owner_name) }}
+    </div>
+
+    <label>{{ 'adyenDirectDebit.iban_number' | trans }}</label>
+    <div class="field">
+        {{ form_widget(form.adyenDirectDebit.iban_number, { 'attr': {'placeholder': 'adyenDirectDebit.iban_number' | trans } }) }}
+        {{ form_errors(form.adyenDirectDebit.iban_number) }}
+    </div>
+</div>
```
Comment (limeeugenia) [inline src/SprykerEco/Yves/Adyen/Theme/default/direct_debit.twig]: newline, please check all twigs you have for this.

File: src/SprykerEco/Yves/Adyen/Theme/default/prepayment.twig [hunk 3]
```diff
@@ -0,0 +1,3 @@
+<div class="adyen-payment-form adyen-payment-sofort-form payment-form" id="adyenPrepayment">
```
Comment (limeeugenia) [inline src/SprykerEco/Yves/Adyen/Theme/default/prepayment.twig]: This is not sofort template.

File: src/SprykerEco/Yves/Adyen/Theme/default/credit_card.twig [hunk 4]
```diff
@@ -1,19 +1,19 @@
...
-        <br />
-        <label>
-            <span class="input-field" data-hosted-id="hostedExpiryDateField" data-cse="encryptedExpiryDate" />
-        </label>
-        <br />
-        <label>
-            <span class="input-field" data-hosted-id="hostedSecurityCodeField" data-cse="encryptedSecurityCode" />
-        </label>
+
+<div class="adyen-payment-form adyen-payment-klarna-invoice-form payment-form cards-div js-chckt-pm__pm-holder" id="adyenKlarnaInvoice">
```
Comment (limeeugenia) [inline src/SprykerEco/Yves/Adyen/Theme/default/credit_card.twig]: This is not a klarna invoice form.

File: src/SprykerEco/Zed/Adyen/Business/Handler/Notification/AdyenNotificationHandler.php [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+        $statuses = $this->config->getMappedOmsStatuses();
+        if (!array_key_exists($notificationTransfer->getEventCode(), $statuses)) {
+            throw new AdyenNotificationNotHandledException('Event code does not registered.');
+        }
+
+        $paymentAdyenTransfer = $this->reader->getPaymentAdyenByPspReference($notificationTransfer->getPspReference());
+        if (!$paymentAdyenTransfer->getFkSalesOrder()) {
+            throw new AdyenNotificationNotHandledException(
+                'Payment entity with requested pspReference does not registered.'
```
Comment (limeeugenia) [inline src/SprykerEco/Zed/Adyen/Business/Handler/Notification/AdyenNotificationHandler.php]: 'Payment entity with requested pspReference **IS** not registered.'

You want a const for this.

File: src/SprykerEco/Zed/Adyen/Business/Handler/Notification/AdyenNotificationHandler.php [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+     */
+    protected function handleNotification(AdyenNotificationRequestItemTransfer $notificationTransfer): void
+    {
+        if (!$notificationTransfer->getSuccess()) {
+            throw new AdyenNotificationNotHandledException('Notification Request Item is not success.');
+        }
+
+        $statuses = $this->config->getMappedOmsStatuses();
+        if (!array_key_exists($notificationTransfer->getEventCode(), $statuses)) {
+            throw new AdyenNotificationNotHandledException('Event code does not registered.');
```
Comment (limeeugenia) [inline src/SprykerEco/Zed/Adyen/Business/Handler/Notification/AdyenNotificationHandler.php]: Event code **IS** not registered.

A const or smthing would be great.

File: src/SprykerEco/Zed/Adyen/Business/Handler/Notification/AdyenNotificationHandler.php [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+     * @param \Generated\Shared\Transfer\AdyenNotificationRequestItemTransfer $notificationTransfer
+     *
+     * @throws \SprykerEco\Zed\Adyen\Business\Exception\AdyenNotificationNotHandledException
+     *
+     * @return void
+     */
+    protected function handleNotification(AdyenNotificationRequestItemTransfer $notificationTransfer): void
+    {
+        if (!$notificationTransfer->getSuccess()) {
+            throw new AdyenNotificationNotHandledException('Notification Request Item is not success.');
```
Comment (limeeugenia) [inline src/SprykerEco/Zed/Adyen/Business/Handler/Notification/AdyenNotificationHandler.php]: Use const for the magic string.

File: src/SprykerEco/Zed/Adyen/Business/Handler/Notification/AdyenNotificationHandler.php [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+     * @param \Generated\Shared\Transfer\AdyenNotificationsTransfer $notificationsTransfer
+     *
+     * @return \Generated\Shared\Transfer\AdyenNotificationsTransfer
+     */
+    public function handle(AdyenNotificationsTransfer $notificationsTransfer): AdyenNotificationsTransfer
+    {
+        $isSuccess = true;
+
+        foreach ($notificationsTransfer->getNotificationItems() as $notificationItem) {
+            try {
```
Comment (limeeugenia) [inline src/SprykerEco/Zed/Adyen/Business/Handler/Notification/AdyenNotificationHandler.php]: It is wrong to use exception for controlling the execution flow. Maybe return boolean from `handleNotification` to find if the message has been saved successfully.

File: src/SprykerEco/Zed/Adyen/Business/Hook/Saver/MakePayment/AliPaySaver.php [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+/**
+ * MIT License
+ * For full license information, please view the LICENSE file that was distributed with this source code.
+ */
+
+namespace SprykerEco\Zed\Adyen\Business\Hook\Saver\MakePayment;
+
+class AliPaySaver extends AbstractSaver
+{
+    protected const MAKE_PAYMENT_PAY_PAL_REQUEST_TYPE = 'MakePayment[AliPay]';
```
Comment (limeeugenia) [inline src/SprykerEco/Zed/Adyen/Business/Hook/Saver/MakePayment/AliPaySaver.php]: Are you sure about the const name?

File: src/SprykerEco/Zed/Adyen/Business/Hook/Saver/MakePayment/WeChatPaySaver.php [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+/**
+ * MIT License
+ * For full license information, please view the LICENSE file that was distributed with this source code.
+ */
+
+namespace SprykerEco\Zed\Adyen\Business\Hook\Saver\MakePayment;
+
+class WeChatPaySaver extends AbstractSaver
+{
+    protected const MAKE_PAYMENT_PAY_PAL_REQUEST_TYPE = 'MakePayment[WeChatPay]';
```
Comment (limeeugenia) [inline src/SprykerEco/Zed/Adyen/Business/Hook/Saver/MakePayment/WeChatPaySaver.php]: Again the const name might be not what you want.

File: src/SprykerEco/Client/Adyen/Zed/AdyenStub.php [hunk 11]
```diff
@@ -2,13 +2,79 @@
...
+    }
+
+    /**
+     * @param \Generated\Shared\Transfer\AdyenNotificationsTransfer $notificationsTransfer
+     *
+     * @return \Generated\Shared\Transfer\AdyenNotificationsTransfer
+     */
+    public function handleNotificationRequest(AdyenNotificationsTransfer $notificationsTransfer): AdyenNotificationsTransfer
+    {
+        /** @var \Generated\Shared\Transfer\AdyenNotificationsTransfer $notificationsTransfer */
```
[the remark this answers is not in the document]
Comment (aleksey-kotsuba) [reply src/SprykerEco/Client/Adyen/Zed/AdyenStub.php]: It's added for cover return type of `->call()` method. Phpstan does not pass without it.

File: src/SprykerEco/Yves/Adyen/Form/IdealSubForm.php [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+     * @param \Symfony\Component\Form\FormBuilderInterface $builder
+     * @param array $options
+     *
+     * @return void
+     */
+    public function buildForm(FormBuilderInterface $builder, array $options): void
+    {
+        parent::buildForm($builder, $options);
+
+        $this->addIdealIssuerField($builder, $options[static::OPTIONS_FIELD_NAME][static::OPTIONS_IDEAL_ISSUER]);
```
[the remark this answers is not in the document]
Comment (aleksey-kotsuba) [reply src/SprykerEco/Yves/Adyen/Form/IdealSubForm.php]: This key always exists.

File: src/SprykerEco/Yves/Adyen/Form/KlarnaInvoiceSubForm.php [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+     * @param \Symfony\Component\Form\FormBuilderInterface $builder
+     * @param array $options
+     *
+     * @return void
+     */
+    public function buildForm(FormBuilderInterface $builder, array $options): void
+    {
+        parent::buildForm($builder, $options);
+
+        if ($options[static::SELECT_OPTIONS_FIELD][KlarnaInvoiceFormDataProvider::SOCIAL_SECURITY_NUMBER_REQUIRED]) {
```
[the remark this answers is not in the document]
Comment (aleksey-kotsuba) [reply src/SprykerEco/Yves/Adyen/Form/KlarnaInvoiceSubForm.php]: This key always exists.

File: src/SprykerEco/Yves/Adyen/Form/WeChatPaySubForm.php [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+     *
+     * @return void
+     */
+    public function configureOptions(OptionsResolver $resolver): void
+    {
+        $resolver->setDefaults([
+            'data_class' => AdyenWeChatPayPaymentTransfer::class,
+        ])->setRequired(static::OPTIONS_FIELD_NAME);
+    }
+}
```
[the remark this answers is not in the document]
Comment (aleksey-kotsuba) [reply src/SprykerEco/Yves/Adyen/Form/WeChatPaySubForm.php]: The keyword is almost :)

File: src/SprykerEco/Yves/Adyen/Theme/default/prepayment.twig (hunk 3 above)
Comment (aleksey-kotsuba) [reply src/SprykerEco/Yves/Adyen/Theme/default/prepayment.twig]: That was adde by our awesome frontend developer @alex-galych. I think he will create an issue and assign it to responsible person.

File: src/SprykerEco/Client/Adyen/AdyenClientInterface.php [hunk 15]
```diff
@@ -2,11 +2,66 @@
...
+     * @param \Generated\Shared\Transfer\AdyenRedirectResponseTransfer $redirectResponseTransfer
+     *
+     * @return \Generated\Shared\Transfer\AdyenRedirectResponseTransfer
+     */
+    public function handleCreditCard3dResponseFromAdyen(AdyenRedirectResponseTransfer $redirectResponseTransfer): AdyenRedirectResponseTransfer;
+
+    /**
+     * Specification:
+     * - Handle response from Adyen after redirect customer back to the shop.
+     * - Save response, update status etc.
```
[the remark this answers is not in the document]
Comment (aleksey-kotsuba) [reply src/SprykerEco/Client/Adyen/AdyenClientInterface.php]: Was fixed.

File: src/SprykerEco/Client/Adyen/AdyenClient.php [hunk 16]
```diff
@@ -2,18 +2,85 @@
...
+    /**
+     * {@inheritdoc}
+     *
+     * @api
+     *
+     * @param \Generated\Shared\Transfer\AdyenNotificationsTransfer $notificationsTransfer
+     *
+     * @return \Generated\Shared\Transfer\AdyenNotificationsTransfer
+     */
+    public function handleNotificationRequest(AdyenNotificationsTransfer $notificationsTransfer): AdyenNotificationsTransfer
```
[the remark this answers is not in the document]
Comment (aleksey-kotsuba) [reply src/SprykerEco/Client/Adyen/AdyenClient.php]: I would like to not use this `FromAdyen` anywhere, but I have to in some cases.

**Pull request closed without merging**